#include "gamewindow.hpp"
#include "ui_gamewindow.h"

#include <QMessageBox>
#include <QScreen>

const QMap<int, QString> GameWindow::NUMBER_COLORS = {
    {1, "blue"}, {2, "green"}, {3, "red"},
    {4, "darkblue"}, {5, "darkred"}, {6, "cyan"},
    {7, "black"}, {8, "gray"}
};

const QString GameWindow::CLOSED_CELL_STYLE =
    "QPushButton {"
    "background: #c0c0c0;"
    "border-top: 1px solid white;"
    "border-left: 1px solid white;"
    "border-right: 1px solid black;"
    "border-bottom: 1px solid black;"
    "}"
    "QPushButton:pressed {"
    "border: 1px inset gray;"
    "background: #b0b0b0;"
    "}";

const QString GameWindow::OPENED_CELL_STYLE =
    "background: #e0e0e0; border: 1px solid gray;";

const QString GameWindow::MINE_CELL_STYLE =
    "background: red; border: 1px solid darkred;";

GameWindow::GameWindow(const Difficulty& currentDifficulty,
                       const Settings& settings,
                       std::shared_ptr<Statistics> statistics,
                       QWidget* parent)

    : QMainWindow(parent),
      ui(std::make_unique<Ui::GameWindow>()),
      game(nullptr)
{

    ui->setupUi(this);

    game = std::make_unique<Game>(currentDifficulty, settings, statistics); // –°–æ–∑–¥–∞—ë–º –∏–≥—Ä—É

    setupUI(currentDifficulty); // –£—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ–º –∏–Ω—Ç–µ—Ä—Ñ–µ–π—Å

    createGameField(); // –°–æ–∑–¥–∞—ë–º –∏–≥—Ä–æ–≤–æ–µ –ø–æ–ª–µ

    setupConnections(); // –ü–æ–¥–∫–ª—é—á–∞–µ–º —Å–∏–≥–Ω–∞–ª—ã

}

GameWindow::~GameWindow() = default;

// –£—Å—Ç–∞–Ω–æ–≤–∫–∞ —Å–∏–≥–Ω–∞–ª–æ–≤ //
void GameWindow::setupConnections() {

    // –°–æ–µ–¥–∏–Ω—è–µ–º –æ–±–Ω–æ–≤–ª–µ–Ω–∏–µ —Ç–∞–π–º–µ—Ä–∞ —Å —Ç–∞–π–º–µ—Ä–æ–º –≤ –æ–∫–Ω–µ //
    connect(&game->getTimer(), &GameTimer::timeUpdated, [this](const QString& time) {
        ui->timerLabel->setText(time);
    });

    // –°–æ–µ–¥–∏–Ω—è–µ–º –∫–Ω–æ–ø–∫—É —Ä–µ—Å—Ç–∞—Ä—Ç–∞ //
    connect(ui->restartTrigger, &QAction::triggered,  ui->restartButton, &QPushButton::click);

    // –°–æ–µ–¥–∏–Ω—è–µ–º —Å–º–µ–Ω—É —Å–ª–æ–∂–Ω–æ—Å—Ç–∏ //
    connect(ui->beginnerDifficulty, &QAction::triggered, this, [this]() {
        changeDifficulty("Beginner", 9, 9, 10);
    });

    connect(ui->intermediateDifficulty, &QAction::triggered, this, [this]() {
        changeDifficulty("Intermediate", 16, 16, 30);
    });

    connect(ui->expertDifficulty, &QAction::triggered, this, [this]() {
        changeDifficulty("Expert", 30, 16, 99);
    });

}

// –£—Å—Ç–∞–Ω–æ–≤–∫–∞ –∏–Ω—Ç–µ—Ä—Ñ–µ–π—Å–∞ //
void GameWindow::setupUI(const Difficulty& currentDifficulty) {

    // –£—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ–º —Å–º–∞–π–ª–∏–∫ –Ω–∞ –∫–Ω–æ–ø–∫—É —Ä–µ—Å—Ç–∞—Ä—Ç–∞ //
    ui->restartButton->setIcon(QIcon(":/images/smile1.svg"));
    ui->restartButton->setIconSize(QSize(64, 64));
    ui->restartButton->setText("");
    ui->restartButton->setFixedSize(70, 70);

    // –£—Å—Ç–∞–Ω–∞–ª–∏–≤–∞–µ–º –∫–æ–ª-–≤–æ –º–∏–Ω –Ω–∞ –ø–æ–ª–µ –∏ —Å–±—Ä–∞—Å—ã–≤–∞–µ–º —Ç–∞–π–º–µ—Ä //
    int mines = currentDifficulty.getMines();
    ui->minesLabel->setText(QString("%1").arg(mines, 3, 10, QChar('0')));
    ui->timerLabel->setText("000");

    setWindowTitle("–°–∞–ø—ë—Ä - " + currentDifficulty.getName());

}

// –°–æ–∑–¥–∞–Ω–∏–µ –∏–≥—Ä–æ–≤–æ–≥–æ –ø–æ–ª—è //
void GameWindow::createGameField() {

    int width = game->getGameField().getWidth();
    int height = game->getGameField().getHeight();

    const int cellSize = calculateOptimalCellSize(width, height);

    // –°–æ–∑–¥–∞—ë–º –º–∞—Å—Å–∏–≤ –∫–Ω–æ–ø–æ–∫ //
    buttons.resize(height);
    for (int y = 0; y < height; y++) {
        buttons[y].resize(width);
    }

    // –°–æ–∑–¥–∞—ë–º Layout //
    QGridLayout* grid = new QGridLayout;
    grid->setSpacing(CELL_SPACING);
    grid->setContentsMargins(0, 0, 0, 0);
    grid->setAlignment(Qt::AlignCenter);

    for (int y = 0; y < height; y++) {
        for (int x = 0; x < width; x++) {

            QPointer<QPushButton> btn = new QPushButton();

            btn->setFixedSize(cellSize, cellSize);
            btn->setSizePolicy(QSizePolicy::Fixed, QSizePolicy::Fixed);

            btn->setStyleSheet(CLOSED_CELL_STYLE);

            buttons[y][x] = btn; // –°–æ—Ö—Ä–∞–Ω—è–µ–º —É–∫–∞–∑–∞—Ç–µ–ª—å –≤ –º–∞—Å—Å–∏–≤

            // –õ–µ–≤—ã–π –∫–ª–∏–∫ –ø–æ –∫–ª–µ—Ç–∫–µ //
            connect(btn, &QPushButton::clicked,
                    [this, x, y]() { onLeftClick(x, y); });

            // –ü—Ä–∞–≤—ã–π –∫–ª–∏–∫ –ø–æ –∫–ª–µ—Ç–∫–µ //
            btn->setContextMenuPolicy(Qt::CustomContextMenu);
            connect(btn, &QPushButton::customContextMenuRequested,
                    [this, x, y]() { onRightClick(x, y); });

            grid->addWidget(btn, y, x); // –î–æ–±–∞–≤–ª—è–µ–º –≤ Layout

        }
    }

    ui->gameArea->setLayout(grid); // –†–∞–∑–º–µ—â–∞–µ–º –ø–æ–ª–µ –≤ –æ–∫–Ω–µ

    setupWindowSize(width, height, cellSize); // –£—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ–º —Ä–∞–∑–º–µ—Ä –æ–∫–Ω–∞

}

// –í—ã—á–∏—Å–ª–µ–Ω–∏–µ –æ–ø—Ç–∏–º–∞–ª—å–Ω–æ–≥–æ —Ä–∞–∑–º–µ—Ä–∞ –∫–ª–µ—Ç–∫–∏ //
int GameWindow::calculateOptimalCellSize(int width, int height) const {

    constexpr int MIN_CELL_SIZE = 20;
    constexpr int MEDIUM_CELL_SIZE = 25;
    constexpr int MAX_CELL_SIZE = 30;

    constexpr int LARGE_FIELD = 25;
    constexpr int HUGE_FIELD = 30;

    int cellSize = MAX_CELL_SIZE;

    if (width > HUGE_FIELD || height > HUGE_FIELD) {
        cellSize = MIN_CELL_SIZE;
    }
    else if (width > LARGE_FIELD || height > LARGE_FIELD) {
        cellSize = MEDIUM_CELL_SIZE;
    }

    return cellSize;
}

// –ò–∑–º–µ–Ω–µ–Ω–∏–µ —Ä–∞–∑–º–µ—Ä–∞ —ç–∫—Ä–∞–Ω–∞ –≤ –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç–∏ –æ—Ç —Å–ª–æ–∂–Ω–æ—Å—Ç–∏ //
void GameWindow::setupWindowSize(int width, int height, int cellSize) {

    int fieldWidth = width * cellSize + (width - 1) * CELL_SPACING;
    int fieldHeight = height * cellSize + (height - 1) * CELL_SPACING;

    setMinimumSize(0, 0);
    setMaximumSize(QWIDGETSIZE_MAX, QWIDGETSIZE_MAX);

    ui->gameArea->setMinimumSize(fieldWidth, fieldHeight);

    adjustSize();

    // –¶–µ–Ω—Ç—Ä–∏—Ä—É–µ–º –æ–∫–Ω–æ //
    QScreen* screen = QGuiApplication::primaryScreen();
    if (screen) {
        QRect screenGeometry = screen->availableGeometry();
        move(screenGeometry.center() - rect().center());
    }

}

// –ù–∞–∂–∞—Ç–∏–µ –ª–µ–≤–æ–π –∫–Ω–æ–ø–∫–æ–π –º—ã—à–∏ //
void GameWindow::onLeftClick(int x, int y) {

    if (!game) return;

    qDebug() << "–õ–µ–≤—ã–π –∫–ª–∏–∫:" << x << y;

    if (game->getGameState() == GameState::Waiting) {
        game->startGame(Point(x, y));
    } else {
        game->cellClick(Point(x, y));
    }

    updateField();

    if (game->getGameState() == GameState::Won) {
        QMessageBox::information(this, "–ü–æ–±–µ–¥–∞!", "–í—ã –≤—ã–∏–≥—Ä–∞–ª–∏!");
    } else if (game->getGameState() == GameState::Lost) {
        QMessageBox::information(this, "–ü–æ—Ä–∞–∂–µ–Ω–∏–µ", "–ú–∏–Ω—ã!");
        game->getGameField().revealAllMines();
        updateField();
    }

}

// –ù–∞–∂–∞—Ç–∏–µ –ø—Ä–∞–≤–æ–π –∫–Ω–æ–ø–∫–æ–π –º—ã—à–∏ //
void GameWindow::onRightClick(int x, int y) {

    if (!game || game->getGameState() != GameState::Running) return;

    qDebug() << "–ü—Ä–∞–≤—ã–π –∫–ª–∏–∫:" << x << y;
    game->flagToggle(Point(x, y));
    updateCell(x, y);

}

// –ù–∞–∂–∞—Ç–∏–µ –Ω–∞ –∫–Ω–æ–ø–∫—É —Ä–µ—Å—Ç–∞—Ä—Ç–∞ //
void GameWindow::on_restartButton_clicked() {

    if (!game) return;

    game->restartGame();

    updateField();

    ui->timerLabel->setText("000");
    int mines = game->getCurrentDifficulty().getMines();
    ui->minesLabel->setText(QString("%1").arg(mines, 3, 10, QChar('0')));
}

// –°–º–µ–Ω–∏—Ç—å —Å–ª–æ–∂–Ω–æ—Å—Ç—å –≤–æ –≤—Ä–µ–º—è –∏–≥—Ä—ã //
void GameWindow::changeDifficulty(QString name, int width, int height, int mines) {

    if (!game) return;

    auto settings = game->getCurrentSettings();
    auto statistics = game->getStatistics();

    // –û—á–∏—â–∞–µ–º —Å—Ç–∞—Ä–æ–µ –ø–æ–ª–µ –≤ UI //
    if (ui->gameArea->layout()) {
        QLayout* oldLayout = ui->gameArea->layout();
        QLayoutItem* item;
        while ((item = oldLayout->takeAt(0)) != nullptr) {
            delete item->widget();
            delete item;
        }
        delete oldLayout;
    }
    buttons.clear();

    game.reset();

    Difficulty newDiff(name, width, height, mines);
    game = std::make_unique<Game>(newDiff, settings, statistics);

    connect(&game->getTimer(), &GameTimer::timeUpdated, [this](const QString& time) {
        ui->timerLabel->setText(time);
    }); // –ü–µ—Ä–µ–ø–æ–¥–∫–ª—é—á–∞–µ–º —Ç–∞–π–º–µ—Ä

    createGameField();

    adjustSize();
    ui->minesLabel->setText(QString("%1").arg(mines, 3, 10, QChar('0')));
    ui->timerLabel->setText("000");
    setWindowTitle("–°–∞–ø—ë—Ä - " + name);

}

// –û–±–Ω–æ–≤–ª–µ–Ω–∏–µ –∏–≥—Ä–æ–≤–æ–≥–æ –ø–æ–ª—è //
void GameWindow::updateField()
{
    if (!game) return;

    const auto& field = game->getGameField();
    for (int y = 0; y < field.getHeight(); ++y) {
        for (int x = 0; x < field.getWidth(); ++x) {
            updateCell(x, y);
        }
    }
}

// –û–±–Ω–æ–≤–ª–µ–Ω–∏–µ –∫–ª–µ—Ç–∫–∏ //
void GameWindow::updateCell(int x, int y) {

    auto btn = getButtonAt(x, y);
    if (!btn || !game) return;

    const Cell* cell = game->getGameField().getCell(x, y);
    if (!cell) return;

    // –°–±—Ä–∞—Å—ã–≤–∞–µ–º —Å–æ–¥–µ—Ä–∂–∞–Ω–∏–µ //
    btn->setIcon(QIcon());
    btn->setText("");

    if (cell->getIsRevealed()) { // –û—Ç–∫—Ä—ã—Ç–∞—è –∫–ª–µ—Ç–∫–∞

        btn->setEnabled(false);

        if (cell->getIsMine()) { // –ï—Å–ª–∏ –∫–ª–µ—Ç–∫–∞ - –º–∏–Ω–∞

            btn->setIcon(QIcon(":/images/mine.svg"));
            btn->setIconSize(QSize(20, 20));
            btn->setStyleSheet(MINE_CELL_STYLE);

        }
        else if (cell->getAdjacentMines() > 0) {

            int m = cell->getAdjacentMines();
            btn->setText(QString::number(m));
            QString color = NUMBER_COLORS.value(m, "black");
            btn->setStyleSheet(OPENED_CELL_STYLE +
                               QString(" color: %1; font-weight: bold;").arg(color));

        }
        else {
            btn->setStyleSheet(OPENED_CELL_STYLE);
        }
    }
    else { // –ó–∞–∫—Ä—ã—Ç–∞—è –∫–ª–µ—Ç–∫–∞

        btn->setEnabled(true);

        if (cell->getIsFlagged()) {
            btn->setText("üö©");
            btn->setStyleSheet(CLOSED_CELL_STYLE + " color: red; font-size: 14px;");
        }
        else {
            btn->setStyleSheet(CLOSED_CELL_STYLE);
        }
    }
}

// –ì–µ—Ç—Ç–µ—Ä –¥–ª—è –∫–Ω–æ–ø–∫–∏ (UI) //
QPushButton* GameWindow::getButtonAt(int x, int y)
{
    if (y >= 0 && y < buttons.size() && x >= 0 && x < buttons[y].size()) {
        return buttons[y][x];
    }
    return nullptr;
}

// –°–æ–±—ã—Ç–∏–µ –ø—Ä–∏ –∑–∞–∫—Ä—ã—Ç–∏–∏ –æ–∫–Ω–∞ //
void GameWindow::closeEvent(QCloseEvent* event)
{
    emit windowClosed(); // –£–≤–µ–¥–æ–º–ª—è–µ–º MainWindow –æ –∑–∞–∫—Ä—ã—Ç–∏–∏ –æ–∫–Ω–∞
    QMainWindow::closeEvent(event); // –ó–∞–∫—Ä—ã–≤–∞–µ–º –æ–∫–Ω–æ
}
